test/object-store/sync/flx_migration.cpp

```diff
@@ -0,0 +1,200 @@
+    for (int i = 0; i < count; i++, ++start) {
+        auto id = ObjectId::gen();
+        auto obj = Object::create(c, realm, "Dog",
+                                  std::any(AnyDict{{"_id", std::any(id)},
+                                                   {"breed", util::format("breed-%1", start)},
+                                                   {"name", util::format("dog-%1", start)},
+                                                   {"realm_id", partition}}));
+        ret.push_back(id);
+    }
+    realm->commit_transaction();
+    return ret;
+}
+
+TEST_CASE("Test server migration and rollback", "[flx],[migration]") {
+    std::shared_ptr<util::Logger> logger_ptr =
+        std::make_shared<util::StderrLogger>(realm::util::Logger::Level::TEST_LOGGING_LEVEL);
+
+    const std::string base_url = get_base_url();
+    const std::string partition1 = "migration-test";
+    const std::string partition2 = "another-value";
+    auto server_app_config = default_app_config(base_url);
+    TestAppSession session(create_app(server_app_config));
+    SyncTestFile config1(session.app(), partition1, server_app_config.schema);
+    SyncTestFile config2(session.app(), partition2, server_app_config.schema);
+
+    // Fill some objects
+    auto objects1 = fill_test_data(config1, partition1, 1, 5);
+    auto objects2 = fill_test_data(config2, partition2, 6, 5);
+
+    auto check_data = [&](SharedRealm& realm, bool check_set1, bool check_set2) {
+        auto table = realm->read_group().get_table("class_Dog");
+        auto partition_col = table->get_column_key("realm_id");
+        auto breed_col = table->get_column_key("breed");
+
+        size_t table_size = [check_set1, check_set2] {
+            if (check_set1 && check_set2)
+                return 10;
+            if (check_set1 || check_set2)
+                return 5;
+            return 0;
+        }();
+
+        REQUIRE(table->size() == table_size);
+        REQUIRE(bool(table->find_first(partition_col, StringData(partition1))) == check_set1);
+        REQUIRE(bool(table->find_first(breed_col, StringData("breed-5"))) == check_set1);
+        REQUIRE(bool(table->find_first(partition_col, StringData(partition2))) == check_set2);
+        REQUIRE(bool(table->find_first(breed_col, StringData("breed-6"))) == check_set2);
+    };
+
+    // Wait for the two partition sets to upload
+    {
+        auto realm1 = Realm::get_shared_realm(config1);
+
+        CHECK(!wait_for_upload(*realm1));
+        CHECK(!wait_for_download(*realm1));
+
+        check_data(realm1, true, false);
+
+        auto realm2 = Realm::get_shared_realm(config2);
+
+        CHECK(!wait_for_upload(*realm2));
+        CHECK(!wait_for_download(*realm2));
+
+        check_data(realm2, false, true);
+    }
+
+    // Migrate to FLX
+    trigger_server_migration(session.app_session(), true, logger_ptr);
+
+    {
+        SyncTestFile flx_config(session.app()->current_user(), server_app_config.schema,
```

Comment:
This will clear the migration data when https://github.com/realm/realm-core/pull/6355 is merged.

Side note: when we add the other tests, we should have one when the user connects as PBS since that's what they would be doing.

Reply:
got it